index.html

```diff
@@ -1,1688 +1,2240 @@
+        foot" user interfaces (presentations, media viewers).
+      </p>
+      <p>
+        Currently, the only way for a gamepad to be used as input would be to
+        emulate mouse or keyboard events, however this would lose information
+        and require additional software outside of the [=user agent=] to
+        accomplish emulation.
+      </p>
+      <p>
+        Meanwhile, native applications are capable of accessing these devices
+        via system APIs.
+      </p>
+      <p>
+        The Gamepad API provides a solution to this problem by specifying
+        interfaces that allow web applications to directly act on gamepad data.
+      </p>
+    </section>
+    <section>
+      <h2>
+        Scope
+      </h2>
+      <p>
+        Interfacing with external devices designed to control games has the
+        potential to become large and intractable if approached in full
+        generality. In this specification we explicitly choose to narrow scope
+        to provide a useful subset of functionality that can be widely
+        implemented and broadly useful.
+      </p>
+      <p>
+        Specifically, we choose to only support the functionality required to
+        support gamepads. Support for gamepads requires two input types:
+        buttons and axes. Both buttons and axes are reported as analog values,
+        buttons ranging from [0..1], and axes ranging from [-1..1].
+      </p>
+      <p>
+        While the primary goal is support for gamepad devices, supporting these
+        two types of analog inputs allows support for other similar devices
+        common to current gaming systems including joysticks, driving wheels,
+        pedals, and accelerometers. As such, the name "gamepad" is exemplary
+        rather than trying to be a generic name for the entire set of devices
+        addressed by this specification.
+      </p>
+      <p>
+        We specifically exclude support for more complex devices that may also
+        be used in some gaming contexts, including those that that do motion
+        sensing, depth sensing, video analysis, gesture recognition, and so on.
+      </p>
+    </section>
+    <section data-dfn-for="Gamepad" data-link-for="Gamepad">
+      <h2>
+        <dfn>Gamepad</dfn> interface
+      </h2>
+      <p>
+        This interface defines an individual gamepad device.
+      </p>
+      <pre class="idl" data-cite="HR-TIME">
+        [Exposed=Window, SecureContext]
+        interface Gamepad {
+          readonly attribute DOMString id;
+          readonly attribute long index;
+          readonly attribute boolean connected;
+          readonly attribute DOMHighResTimeStamp timestamp;
+          readonly attribute GamepadMappingType mapping;
+          readonly attribute FrozenArray&lt;double&gt; axes;
+          readonly attribute FrozenArray&lt;GamepadButton&gt; buttons;
+          [SameObject] readonly attribute GamepadHapticActuator? vibrationActuator;

```

Comment:
To support `pulse()` we probably need to bring back the `hapticActuators` array:

```
readonly attribute FrozenArray&lt;GamepadHapticActuator> hapticActuators;
```

With `pulse()` it's expected that a gamepad could have multiple haptic actuators and each actuator can be controlled independently. With `playEffect()` the `vibrationActuator` member is expected to control multiple actuators. This means there should be overlap between the actuators in `hapticActuators` and `vibrationActuator`. The spec should explain how to handle pre-empting `pulse()` with `playEffect()` and vice versa.

Reply:
When trying to write this, I felt like it is a little bit confusing to have both  a `FrozenArray<GamepadHapticActuator> hapticActuators` and a `GamepadHapticActuator vibrationActuator` on the the Gamepad interface. For example, to keep consistency, if we decide to also expose `hapticActuators`, should the two dual-rumble motors also be individually exposed as 2 extra elements of the `hapticActuators` array? If that is the case, why should we do it when we can achieve the same level of control by calling playEffect? We can activate only one of the motors with:
```js
gamepad.vibrationActuator.playEffect("dual-rumble", {
        duration: 1000,
        strongMagnitude: 0.5,
 });
```

As mentioned in #148, the Quest Controller Pro has 3 haptic actuators (one in the grip, one in the thumb rest, and another in the trigger). I see that the `pulse()` proposal was made some time ago already. Maybe, this type of controller could take advantage of the trigger rumble extension proposed earlier (#138). Another issue that I see with the `hapticActuators` approach is that there is no standardized way (yet) for us to identify the actuators (which one is the left trigger? which one is the right ERM?) and actuate the desired one. 

For now, I would suggest to postpone adding the `hapticActuators` array and the `pulse()` method to the Gamepad spec and, instead, keep `pulse()` in the Gamepad Extensions document inside a partial GamepadHapticActuator Interface. Later on we can discuss further how to support both `hapticActuators` and `pulse()`. What do you think?

Reply:
> to keep consistency, if we decide to also expose hapticActuators, should the two dual-rumble motors also be individually exposed as 2 extra elements of the hapticActuators array?
 
I think we should expose dual-rumble actuators as 2 elements in `hapticActuators` as well as a non-null `vibrationActuator`.

> there is no standardized way (yet) for us to identify the actuators (which one is the left trigger? which one is the right ERM?) and actuate the desired one

The standard can be opinionated on actuator indices like it is for button and axis indices. If we follow the Standard Gamepad example then we should define standard actuators based on layout. That said, we generally can't detect layout programmatically for "dual-rumble" gamepads. APIs usually label them as strong/weak without specifying where they are located. I don't think the standard should care where they are located either since the intent is to vibrate the whole gamepad, not one particular side of it. Still, we need some way for apps to distinguish them.

I would be fine with using the same indices that Android uses (index 0 for strong vibration channel, index 1 for weak vibration channel). I think it makes sense to give these special placement since they are the most common type of gamepad haptic effect.

For gamepads with actuators that aren't intended to vibrate the entire gamepad (like Xbox trigger rumble or Quest Controller Pro) we should probably start thinking about adding GamepadHapticActuator attributes to identify placement instead of relying on standard indices. Perhaps we can lean on OpenXR's concept of semantic paths to describe actuator placement in a generalizable way.

https://registry.khronos.org/OpenXR/specs/1.0/html/xrspec.html#semantic-path

> why should we do it when we can achieve the same level of control by calling playEffect?

`pulse()` should not preempt effects playing on other actuators. With `playEffect()` apps need to issue a new effect each time either actuator's vibration should change. This is fine for dual-rumble but I think it might be cumbersome for trigger-rumble and other actuator configurations.

> For now, I would suggest to postpone adding the hapticActuators array and the pulse() method to the Gamepad spec and, instead, keep pulse() in the Gamepad Extensions document inside a partial GamepadHapticActuator Interface. Later on we can discuss further how to support both hapticActuators and pulse(). What do you think?

That's fine with me, we can focus on merging `vibrationActuator` and `playEffect()` for now to unblock trigger-rumble. There aren't any implementations yet that implement both `vibrationActuator` and `hapticActuators` so it's probably premature to define their interaction in the main spec. If we keep `hapticActuators` and `pulse()` in the extensions doc then we can iterate on them there.

Reply:
Great! Thanks for your always super detailed answers! I have created a partial GamepadHapticActuator interface in the extensions document.